File: python3/279PerfectSquares.py
class Solution:
    def numSquares(self, n):
        """
        :type n: int
        :rtype: int
        """
        array = [0] * (n + 1)
        
        idx = 1
        while idx < len(array):
            index = 1
            array[idx] = n
            while index * index <= idx:
                if array[idx - index * index] + 1 < array[idx]:
                    array[idx] = array[idx - index * index] + 1
                index += 1
            idx += 1
        return array[n]

File: java/279PerfectSquares.java
class Solution {
    public int numSquares(int n) {
        int[] array = new int[n+1];
        
        for (int i = 1; i <= n; ++i) {
            array[i] = n;
            for (int j = 1; j * j <= i; ++j) {
                if (array[i - j * j] + 1 < array[i])
                    array[i] = array[i - j * j] + 1;
            }
        }
        
        return array[n];
    }
}
